pkg/volume/util/atomic_writer.go

```diff
@@ -0,0 +1,305 @@
+// target directory that should hold the current projected state.
+func (w *AtomicWriter) currentStateDirName() string {
+	return fmt.Sprintf("..%v", w.IdentifierToken)
+}
+
+// updateStateDirName returns the name
+func (w *AtomicWriter) updateStateDirName() string {
+	return fmt.Sprintf("..%v_tmp", w.IdentifierToken)
+}
+
+// Write ensures that the given payload is projected onto disk, updating the
+// contents of the AtomicWriter's target directory atomically if any writes
+// need to be performed.  A sentinel file named "..sentinel" in the target
+// directory is updated, allowing consumers of the directory to listen for
+// updates to the projected data by monitoring the sentinel file with inotify
+// or fanotify.
+//
+// The visible files in this volume are symlinks to files in the writer's
+// current state directory directory.  Actual files are stored in a hidden
+// timestamped directory which is symlinked to by the current state directory.
+// The timestamped directory and current state symlink are created in the
+// writer's target dir.  This scheme allows the files to be atomically updated by
+// changing the target of the current state symlink.
+//
+// When Write is called:
+//
+// 1.  A check is made to determine whether the payload requires writing to disk.
+//     If the payload is already present on disk, the function returns.
+// 2.  A new timestamped dir is created by newTimestampDir
+// 3.  The payload is written to the new timestamped directory
+// 4.  Symlinks and directory for new files are created (if needed).
+//
+//     For example for files:
+//       <target-dir>/user_space/labels
+//       <target-dir>/k8s_space/annotations
+//       <target-dir>/podName
+//
+//     This structure is created:
+//       <target-dir>/podName               -> ..<identifier-token>/podName
+//       <target-dir>/user_space/labels     -> ../..<identifier-token>/user_space/labels
+//       <target-dir>/k8s_space/annotations -> ../..<identifier-token>/k8s_space/annotations
+//       <target-dir>/..<identifier-token>  -> ..<identifier-token>.12345678
+//
+//     where ..<identifier-token>.12345678 is a randomly generated directory that
+//     contains the real data.  If a file has to be dumped in subdirectory
+//     (for example <target-dir>/user_space/labels), the writer builds a
+//     relative symlink (<target-dir>/user_space/labels -> ../..<identifier-token>/user_space/labels)
+// 5.  The previous timestamped directory is detected reading the '..<identifier-token>' symlink
+// 6.  A new temporary symlink is created: ..<identifier-token>_tmp
+// 7.  ..<identifier-token>_tmp is renamed to ..<identifier-token>
+// 8.  The previous timestamped directory is removed, if it exists
+// 9.  The sentinel file modification and access times are updated (file is created if it does not
+//     already exist)
+func (w *AtomicWriter) Write(payload map[string][]byte) error {
+	shouldWritePayload, err := w.shouldWritePayload(payload)
```

Comment:
I will adopt it here, and maybe in a couple other places, but this is basically how i feel about local-scope:
![.](https://seussblog.files.wordpress.com/2013/10/bpibh00065_2.jpg)